### PR TITLE
implement union type support for streaming actions

### LINKED ISSUE #607: Streaming Event type, type hint, should support union type
# Current behavior
When it's a union type it complains:
```
union = MyModel1 | MyModel2

@streaming_action.pydantic(
    reads=[...],
    writes=[...],
    state_input_type=..., 
    state_output_type=...
    stream_type=union, # type hint says this is invalid
)
```

because it is typed with :

```
stream_type: Union[Type["BaseModel"], Type[dict]],
```

## Library & System Information
Latest burr

# Expected behavior
It should be fine.

# Additional context
The fix I think is to expand that stream type to handle a union of BaseModel types.

### PR DESCRIPTION
**Pull request by ThyTran1402**
## Summary
Extends the `@streaming_action.pydantic()` decorator to accept union types like `Model1 | Model2` or `Union[Model1, Model2]`) for the `stream_type` parameter. Previously, only single BaseModel types or `dict` were accepted, limiting flexibility for streaming actions that yield multiple result types.

## Changes
- **burr/core/action.py**
- **burr/integrations/pydantic.py**
- **tests/integrations/test_streaming_union_types.py**

## How I tested this
**Unit Tests**: Created comprehensive test suite with 10 tests that all pass
## Notes
Resolved: #607 

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by Ghraven** [comment]
Thanks for picking this up and for writing out the investigation so clearly.

One small thought from my earlier pass on this issue: adding object to the union does make the type checker accept union values, but it also effectively widens stream_type to almost any value. That may still be the pragmatic choice here, but maintainers may prefer either a more explicit alias/comment around why the parameter is intentionally broad, or a small runtime/type-helper path that recognizes union forms while keeping the public intent narrow.

The focused tests for both 	yping.Union[...] and ModelA | ModelB look useful. Thanks again for moving this forward.

**Comment by ThyTran1402** [comment]
Thank you for your suggestions. That's a good catch. I addressed the issues that you mentioned above.

**Comment by Ghraven** [comment]
Thank you for following up and tightening that path. I appreciate you taking the suggestion constructively, and the updated direction looks much clearer to me. Nice work.

File: burr/integrations/pydantic.py [hunk 1]
```diff
@@ -307,6 +314,10 @@ def _validate_and_extract_signature_types_streaming(
     if stream_type is None:
         # TODO -- derive from the signature
         raise ValueError(f"stream_type is required for function: {fn.__qualname__}")
+    if not _is_valid_stream_type(stream_type):
```
Comment (Haricharanpanjwani) [inline burr/integrations/pydantic.py:317]: This looks like the key behavior change in the PR, so I think we should add negative-path coverage for it as well. Right now the new tests show that valid unions are accepted, but they don’t pin the rejection behavior for invalid cases like `stream_type=int` or `stream_type=Union[TextChunk, int]`. Could we add a couple of `pytest.raises(ValueError)` assertions for those?

File: burr/core/action.py [hunk 2]
```diff
@@ -1507,15 +1507,15 @@ def pydantic(
         writes: List[str],
         state_input_type: Type["BaseModel"],
         state_output_type: Type["BaseModel"],
-        stream_type: Union[Type["BaseModel"], Type[dict]],
+        stream_type: Union[Type["BaseModel"], Type[dict], object],
```
Comment (Haricharanpanjwani) [inline burr/core/action.py]: This fixes the usability issue, but it also broadens the public type hint a lot. Is that tradeoff intentional? We’re now depending on runtime validation for the real constraint, so static tooling no longer gets much help from the signature.

File: tests/integrations/test_streaming_union_types.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    assert hasattr(act, "bind")
+    assert hasattr(act, FunctionBasedAction.ACTION_FUNCTION)
+
+
+# ============================================================================
+# Union Type Tests - typing.Union Syntax
+# ============================================================================
+
+
+def test_streaming_action_typing_union_two_models():
```
Comment (Haricharanpanjwani) [inline tests/integrations/test_streaming_union_types.py:129]: The happy-path coverage here is good, but I think we’re still missing one or two failure-mode tests for the new validator. Adding invalid `stream_type` cases would make the review much easier because it would pin the exact boundary this PR is introducing.

File: burr/integrations/pydantic.py (hunk 1 above)
Comment (ThyTran1402) [reply burr/integrations/pydantic.py:317]: Yes. That's valid. I added more tests for existing positive path and negative path.

File: burr/core/action.py (hunk 2 above)
Comment (ThyTran1402) [reply burr/core/action.py]: Yes. A `typing.Union[...]` value has no precise static type that mypy evaluates the expression as object, so any narrower hint would produce false positives for the exact use case this PR enables and we support 3.9, where pipe syntax isn't available. This is the same pattern FastAPI uses for `response_model:` `Any`. The real constraint is enforced at decoration time by `_is_valid_stream_type,` and the rejection behavior is now pinned by negative-path tests (per your other comment). I've named the type of alias and documented the tradeoff at the definition site, so it reads as deliberate.

File: tests/integrations/test_streaming_union_types.py (hunk 3 above)
Comment (ThyTran1402) [reply tests/integrations/test_streaming_union_types.py:129]: Thank you for your review and feedback. I addressed this.

**Comment by ThyTran1402** [comment]
Hi @andreahlert,

I think the PR is ready for review. Can you please take a look at the PR when you have time?

Thank you!
